### DISCUSSION TITLE
`https://github.com/<org>/<repo>/pull/<number>/commits/<sha>.patch?full_index=1` redirects without `?full_index=1`

### DISCUSSION THREAD

**Discussion by haampie** [community/community#126582] [Repositories]
### Select Topic Area

Bug

### Body

In the https://github.com/spack/spack package manager we store patches using GitHub URLs plus shasums:

```python
patch(
    "https://github.com/protocolbuffers/protobuf/pull/11032/commits/3039f932aaf212bcf2f14a3f2fd00dbfb881e46b.patch?full_index=1",
    sha256="a779238fb7957514d4fb393410111419a964771e826ec2a8f09c21aa1efbb4d1",
)
```

Recently, this particular checksum broke. The reason is a regression in GitHub:

```console
$ curl -fIsS 'https://github.com/protocolbuffers/protobuf/pull/11032/commits/3039f932aaf212bcf2f14a3f2fd00dbfb881e46b.patch?full_index=1' | grep location
location: https://github.com/protocolbuffers/protobuf/commit/3039f932aaf212bcf2f14a3f2fd00dbfb881e46b.patch
```

this redirect is missing the original `?full_index=1` query arg.

The checksum with `?full_index=1` is correct, the one without wrong:

```console
$ curl -Lfs 'https://github.com/protocolbuffers/protobuf/commit/3039f932aaf212bcf2f14a3f2fd00dbfb881e46b.patch?full_index=1' | sha256sum 
a779238fb7957514d4fb393410111419a964771e826ec2a8f09c21aa1efbb4d1  -

$ curl -Lfs 'https://github.com/protocolbuffers/protobuf/commit/3039f932aaf212bcf2f14a3f2fd00dbfb881e46b.patch' | sha256sum 
ae8639a308294bed85d3f6ebb7e005ee4fde6b2884d41cb75543191c9c5c98ef  -
```

It would be great if GitHub responds with the `?full_index=1` bit in the `location:` header when it redirects.

**Comment by wdconinc** [comment]
See https://github.com/spack/spack/pull/44210 and https://github.com/spack/spack/pull/44212 for original discussion about this issue in spack. (@haampie I think your first `patch` example above isn't actually an affected link, though in the second code block you use an affected link.)

**Reply by haampie** [reply]
Thanks, updated the example.
